src/NLog/Internal/MessageTemplateParameters.cs

```diff
@@ -0,0 +1,100 @@
+// 
+// Redistribution and use in source and binary forms, with or without 
+// modification, are permitted provided that the following conditions 
+// are met:
+// 
+// * Redistributions of source code must retain the above copyright notice, 
+//   this list of conditions and the following disclaimer. 
+// 
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution. 
+// 
+// * Neither the name of Jaroslaw Kowalski nor the names of its 
+//   contributors may be used to endorse or promote products derived from this
+//   software without specific prior written permission. 
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE 
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF 
+// THE POSSIBILITY OF SUCH DAMAGE.
+// 
+
+namespace NLog.Internal
+{
+    using System.Collections;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// Internal implementation of the interface for returning MessageTemplate parameters
+    /// </summary>
+    internal class MessageTemplateParameters : IMessageTemplateParameters
+    {
+        readonly IList<MessageTemplateParameter> _parameters;
+
+        /// <inheritDoc/>
+        public MessageTemplateParameter this[int index] { get { return _parameters[index]; } }
+
+        /// <inheritDoc/>
+        public int Count { get { return _parameters.Count; } }
+
+        /// <inheritDoc/>
+        public IEnumerator<MessageTemplateParameter> GetEnumerator() { return _parameters.GetEnumerator(); }
+
+        /// <inheritDoc/>
+        IEnumerator IEnumerable.GetEnumerator() { return _parameters.GetEnumerator(); }
+
+        /// <summary>
+        /// Constructore for positional parameters
+        /// </summary>
+        public MessageTemplateParameters(object[] parameters)
+        {
+            if (parameters != null && parameters.Length > 0)
+            {
+                _parameters = new MessageTemplateParameter[parameters.Length];
+                for (int i = 0; i < parameters.Length; ++i)
+                {
+                    string parameterName;
+                    switch (i)
+                    {
+                        case 0: parameterName = "0"; break;
```

Comment:
this is to prevent boxing, isn't?

Reply:
No to prevent allocation of string-objects.